src/permutation.jl

```diff
@@ -0,0 +1,248 @@
+
+ter Braak C.J.F. (1992) Permutation Versus Bootstrap Significance Tests in
+Multiple Regression and Anova. In: Jöckel KH., Rothe G., Sendler W. (eds)
+Bootstrapping and Related Techniques. Lecture Notes in Economics and Mathematical
+Systems, vol 376. Springer, Berlin, Heidelberg.
+https://doi.org/10.1007/978-3-642-48850-4_10
+
+and
+
+Winkler, A. M., Ridgway, G. R., Webster, M. A., Smith, S. M., & Nichols, T. E. (2014).
+Permutation inference for the general linear model. NeuroImage, 92, 381–397.
+https://doi.org/10.1016/j.neuroimage.2014.01.060
+"""
+function permute!(rng::AbstractRNG, mod::LinearMixedModel{T},
+                   blups=ranef(mod),
+                   reterms=mod.reterms;
+                   β::AbstractVector{T}=coef(mod),
+                   residual_method=:signflip) where {T}
+    y = response(mod) # we are now modifying the model
+    copy!(y, residuals(mod))
+
+    if residual_method == :shuffle
+        shuffle!(rng, y)
+    elseif  residual_method == :signflip
+        y .*= rand(rng, (-1,1), length(y))
+    else
+        throw(ArgumentError("Invalid: residual permutation method: $(residual_method)"))
+    end
+
+    for (re, trm) in zip(blups, reterms)
+        npreds, ngrps = size(re)
+        # sign flipping
+        newre = re * diagm(rand(rng, (-1,1), ngrps))
+
+        # our RE are actually already scaled, but this method (of unscaledre!)
+        # isn't dependent on the scaling (only the RNG methods are)
+        MixedModels.unscaledre!(y, trm, newre)
+    end
+
+    mul!(y, mod.X, β, one(T), one(T))
+
+    # mark model as unfitted
+    mod.optsum.feval = -1
+
+    return mod
+end
+
+
+permutationtest(perm::MixedModelPermutation, model::LinearMixedModel) = permutationtest(perm::MixedModelPermutation, model, :twosided)
+
+"""
+    permutationtest(perm::MixedModelPermutation, model, type=:greater)
+
+Perform a permutation using the already computed permutation and given the observed values.
+
+The `type` parameter specifies use of a two-sided test (`:twosided`) or the directionality of a one-sided test
+(either `:lesser` or `:greater`, depending on the hypothesized difference to the null hypothesis).
+
+See also [`permutation`](@ref).
+
+"""
+function permutationtest(perm::MixedModelPermutation, model, type::Symbol=:twosided)
+    if type == :greater || type  == :twosided
+        comp = >
+    elseif type == :lesser
+        comp = <
+    else
+        throw(ArgumentError("Comparison type $(type) unsupported"))
+    end
+
+    ests = Dict(Symbol(k) => v for (k,v) in zip(coefnames(model), coef(model)))
+    perms = columntable(perm.β)
+
+    dd = Dict{Symbol, Vector}()
+
+    for k in Symbol.(coefnames(model))
+        dd[k] = perms.β[perms.coefname .== k]
+        if type == :twosided
+              μ = mean(dd[k])
+              dd[k] .= abs.(dd[k] .- μ)
+              ests[k] = abs(ests[k].- μ)
+        end
+    end
+    results = (; (k => mean(comp(ests[k]), v) for (k,v) in dd)...)
```

Comment:
Maybe I missed it somewhere, but you have to adjust the pvalue to never be exactly 0 (because observed data belong to permutation set). There are multiple ways to do it
![grafik](https://user-images.githubusercontent.com/10183650/108119294-dcb6a380-709f-11eb-946d-897c884690b9.png)

Obv. the first one is the simplest one, (does it reduce to adding 1/(nperm+1) to your p-value ?)

But isn't there a julia library for permutation tests? That would be even more versatile. I am also asking, because there are several methods to try to estimate the tails of permutation distributions, in order to reduce the number of permutations needed. Maybe those are already implemented there as well

Reply:
moved to a separate issue